### ISSUE TITLE
t.plugin.getCache craps out on Windows with Python 2.5

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2136]
|<img alt="PenguinOfDoom's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| PenguinOfDoom reported|
|-|-|
|Trac ID|trac#2136|
|Type|defect|
|Created|2006-09-30 06:01:40Z|

I'm using patched WinXP and python2.5 (not sure if that's relevant). os.listdir on a non-existent directory fails with ERROR_FILE_NOT_FOUND, which getCache does not expect. Here's a patch:

```
Index: twisted/plugin.py
===================================================================
--- twisted/plugin.py   (revision 18338)
+++ twisted/plugin.py   (working copy)
@@ -118,6 +118,7 @@
         """
 
 # http://msdn.microsoft.com/library/default.asp?url=/library/en-us/debug/base/system_error_codes.asp
+ERROR_FILE_NOT_FOUND = 2
 ERROR_PATH_NOT_FOUND = 3
 ERROR_INVALID_NAME = 123
 
@@ -136,7 +137,7 @@
         try:
             dropinNames = os.listdir(p)
         except WindowsError, e:
-            if e.errno == ERROR_PATH_NOT_FOUND:
+            if e.errno in (ERROR_FILE_NOT_FOUND, ERROR_PATH_NOT_FOUND):
                 continue
             elif e.errno == ERROR_INVALID_NAME:
                 log.msg("Invalid path %r in search path for %s" % (p, module.__name__))
```

Attachments:

* [2136.diff](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ac5/ac5c0955be131109298da1d8cb38a50d16c9e9ba/deac3cec0b25725d4d38c6881171ea3c309d527d.diff) (780 bytes) - added by PenguinOfDoom on 2006-09-30 06:03:58Z - 
* [filepath.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ac5/ac5c0955be131109298da1d8cb38a50d16c9e9ba/b2f57e5366bb404bc35f0c8e58f6827e5040d3a5.py) (15602 bytes) - added by scmikes on 2006-11-06 22:19:09Z - 
* [plugin.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ac5/ac5c0955be131109298da1d8cb38a50d16c9e9ba/c9ec460fc169965920d0ef8d0f660b21771ab584.py) (6691 bytes) - added by scmikes on 2006-11-06 22:19:35Z - 
* [patch.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/ac5/ac5c0955be131109298da1d8cb38a50d16c9e9ba/c9d1be3a12c00667bfdaa35aaaabbe1d8fdcd585.patch) (6011 bytes) - added by scmikes on 2006-11-06 22:25:29Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__2136 2136
type__defect defect
reporter__PenguinOfDoom PenguinOfDoom
priority__highest highest
milestone__Core_2_5 Core-2.5
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__core core
keywords__win32 win32
time__1159596100000000 1159596100000000
changetime__1163560808000000 1163560808000000
version__None None
owner__ 
cc__TimothyFitz cc__teratorn
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_1"></a>|@glyph commented|
|-|-|

This needs a test, and I bet it would be easier to write if we just bit the bullet and moved plugins to use filepath.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun commented|
|-|-|

Python 2.5 is in fact relevant.  Python 2.4 and earlier raised OSError here, not WindowsError.

**Comment by twisted-trac** [comment]
|<img alt="scmikes's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_4"></a>|scmikes commented|
|-|-|

This should be raised to highest, blocker.  This defect is preventing trial from running on the python 2.5 windows buildbot.

**Comment by twisted-trac** [comment]
|<img alt="scmikes's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|scmikes commented|
|-|-|

I don't know how to write a test case for this. but I have been running with the patch locally for several weeks with no problems.

**Comment by twisted-trac** [comment]
|<img alt="scmikes's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_6"></a>|scmikes commented|
|-|-|

This test case validates the python idiom used 
###############################################
    def testPythonDirNotFound(self):
        """ test that exceptions that python throws when files not found 
        
        This test validates that the semantic behavior of detecting a directory that 
        does not exist is correct.
        
        This python idiom is used in plugins.py getcache()
        
        """
        p="zzzThisDirDoesNotExist"
        
        # http://msdn.microsoft.com/library/default.asp?url=/library/en-us/debug/base/system_error_codes.asp
        ERROR_FILE_NOT_FOUND = 2 
        ERROR_PATH_NOT_FOUND = 3
        ERROR_INVALID_NAME = 123
        
        dirNotValid=False
        
        for x in range(2):
                try:
                    dropinNames = os.listdir(p)
                except WindowsError, e:
                    if e.errno in (ERROR_FILE_NOT_FOUND, ERROR_PATH_NOT_FOUND):
                    #  mls old code from plugin ->            if e.errno == ERROR_PATH_NOT_FOUND:
                        dirNotValid=True
                        continue
                    elif e.errno == ERROR_INVALID_NAME:
                        #log.msg("Invalid path %r in search path for %s" % (p, module.__name__))
                        print "invalid path"
                        continue
                    else:
                        raise
                except OSError, ose:
                    if ose.errno not in (errno.ENOENT, errno.ENOTDIR):
                        raise
                    else:
                        continue   
                     
        self.assertTrue(dirNotValid)

**Comment by twisted-trac** [comment]
|<img alt="scmikes's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_7"></a>|scmikes commented|
|-|-|

added 3 braces around code

```


    def testPythonDirNotFound(self):
        """ test that exceptions that python throws when files not found 
        
        This test validates that the semantic behavior of detecting a directory that 
        does not exist is correct.
        
        This python idiom is used in plugins.py getcache()
        
        """
        p="zzzThisDirDoesNotExist"
        
        # http://msdn.microsoft.com/library/default.asp?url=/library/en-us/debug/base/system_error_codes.asp
        ERROR_FILE_NOT_FOUND = 2 
        ERROR_PATH_NOT_FOUND = 3
        ERROR_INVALID_NAME = 123
        
        dirNotValid=False
        
        for x in range(2):
                try:
                    dropinNames = os.listdir(p)
                except WindowsError, e:
                    if e.errno in (ERROR_FILE_NOT_FOUND, ERROR_PATH_NOT_FOUND):
                    #  mls old code from plugin ->            if e.errno == ERROR_PATH_NOT_FOUND:
                        dirNotValid=True
                        continue
                    elif e.errno == ERROR_INVALID_NAME:
                        #log.msg("Invalid path %r in search path for %s" % (p, module.__name__))
                        print "invalid path"
                        continue
                    else:
                        raise
                except OSError, ose:
                    if ose.errno not in (errno.ENOENT, errno.ENOTDIR):
                        raise
                    else:
                        continue   
                     
        self.assertTrue(dirNotValid)

```

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)|@jml set owner to @jml|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_9"></a>|@glyph commented|
|-|-|

I was going to reassign, but for the fact that jml stole this so recently.  Anyway, our illustrious Windows team really ought to hear about such things.

**Comment by twisted-trac** [comment]
|<img alt="scmikes's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|scmikes commented|
|-|-|

I don't know if this is useful, but I started working on the switch to t.p.filepath.

Someone more familiar with getCache can make better use of filepath.

I did make changes to t.p.filepath to throw an exception t.p.filepath.DirNotFound when directory does not exist.

here is a test case for FilePath.saneListdir
```

    def testPythonDirNotFound(self):
        """ test that t.p.DirNotFound is thrown when directory is not found """
        p="zzzThisDirDoesNotExist"
      
        from twisted.python.filepath import FilePath, DirNotFound
        fp = FilePath(p)
        
        self.assertRaises(DirNotFound, fp.saneListdir)
        

I will attach the modified twisted file, use them if they help, ignore the if they do not help



```

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun removed owner|
|-|-|

Probably fixed in plugin-cache-2136.  Cannot test since the Win32 Python 2.5 buildslave is offline at the moment.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_14"></a>|@jml set owner to @exarkun|
|-|-|

The code looks pretty good. I'm not exactly sure how the added tests exercise the changed code -- I guess Windows raises a new kind of error.

Merge away.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun set status to `closed`|
|-|-|

(In [18690]) Merge plugin-cache-2136

Author: exarkun
Reviewer: jml
Fixes [#2136](https://github.com/twisted/twisted/issues/2136)

Add error handling for ERROR_FILE_NOT_FOUND to the plugin system for
Python 2.5 on Windows.  Also add test coverage for the environment
which can result in this error.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
